### PR TITLE
RHCLOUD-46010: Bring Create/Delete/ReadTuples and AcquireLock endpoints over to inventory-api from relations-api

### PR DESCRIPTION
**Pull request by merlante**
### PR Template:

## Describe your changes

  - **Added tuple CRUD endpoints** from relations-api: CreateTuples, DeleteTuples, ReadTuples, AcquireLock (marked DEPRECATED)                                                                             
  - **Proto definitions** with "Relation" prefix for layer separation: `Relation{ObjectType,ObjectReference,SubjectReference,TupleFilter,SubjectFilter,FencingCheck}`, `Relationship`
  - **Use-case layer** (`TupleCrudUsecase`) with business logic, conversion utilities, and meta authorization enforcement                                                                                  
  - **Access control** via `WhitelistMetaAuthorizer` - restricts access to RBAC service only using OIDC ClientID allowlist                                                                                 
  - **Security enforcement**: gRPC-only protocol, OIDC-only authentication, fail-closed by default (empty allowlist denies all)                                                                            
  - **Configuration**: `--metaauthorizer.tuple-crud-allowlist` flag (ConfigMap: `metaauthorizer.tuple-crud-allowlist: ["rbac-service"]`)                                                                   
  - **Tests**: 940+ lines of tests added (unit, integration, conversion utilities) - all passing                                                                                                           
                                                                                                                                                                                                           
  These endpoints are legacy shims for RBAC service migration and designed for easy future removal. 

## Ticket reference (if applicable)
Fixes #RHCLOUD-46010

## Checklist

* [ ] Are the agreed upon acceptance criteria fulfilled?

* [ ] Was the 4-eye-principle applied? (async PR review, pairing, ensembling)

* [ ] Do your changes have passing automated tests and sufficient observability?

* [ ] Are the work steps you introduced repeatable by others, either through automation or documentation?
  * [ ] If automation is possible but not done due to other constraints, a ticket to the tech debt sprint is added
  * [ ] An SOP (Standard Operating Procedure) was created

* [ ] The Changes were automatically built, tested, and  - if needed, behind a feature flag - deployed to our production environment. (**Please check this when the new deployment is done and you could verify it.**)

* [ ] Are the agreed upon coding/architectural practices applied?

* [ ] Are security needs fullfilled? (e.g. no internal URL)

* [ ] Is the corresponding Ticket in the right state? (should be on "review" now, put to done when this change made it to production)

* [ ] For changes to the public API / code dependencies: Was the whole team (or a sufficient amount of ppl) able to review?



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a deprecated tuple-management API (create, delete, read stream, acquire lock) for legacy RBAC compatibility.
  * Added a configurable meta-authorizer with allowlist-based controls to gate tuple operations.

* **Tests**
  * Added extensive tests covering tuple flows, conversions, streaming behavior, and meta-authorizer enforcement.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: api/kessel/inventory/v1beta2/create_tuples_request.proto [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+option java_package = "org.project_kessel.api.inventory.v1beta2";
+
+// DEPRECATED: CreateTuplesRequest is part of the legacy tuple-layer API.
+// This type exists only for RBAC backward compatibility and will be removed.
+message CreateTuplesRequest {
+  // Whether or not the request should ignore existing tuples (true),
+  // or if the request should fail if the same tuple already exists (false).
+  // Defaults to false.
+  bool upsert = 1;
+  repeated Relationship tuples = 2;
```
[the remark this answers is not in the document]
Comment (merlante) [reply api/kessel/inventory/v1beta2/create_tuples_request.proto:19]: Proto is largely inherited from relations-api directly; however, I am addressing some of these proto validation issues identified by coderabbit in this PR, but only when it would not represent a breaking change versus relations-api. In this case, we have tests like 
`  func TestRelationshipsService_CreateRelationshipsWithNilRelationshipsSlice(t *testing.T) {                                                                                                               
        t.Parallel()                                        
        relationshipsService, err := setup(t)                                                                                                                                                              
        assert.NoError(t, err)                                                                                                                                                                             
        ctx := context.Background()                                                                                                                                                                        
                                                                                                                                                                                                           
        req := &v1beta1.CreateTuplesRequest{                                                                                                                                                               
                Tuples: nil,                                                                                                                                                                               
        }                                                                                                                                                                                                  
        _, err = relationshipsService.CreateTuples(ctx, req)
        assert.NoError(t, err)  // <-- EXPECTS NO ERROR
  } `
which indicate that not only do nil/empty arrays get through proto validation, but they are also tolerated internally.

File: api/kessel/inventory/v1beta2/acquire_lock_request.pb.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
+// Code generated by protoc-gen-go. DO NOT EDIT.
```
Comment (snehagunta) [inline api/kessel/inventory/v1beta2/acquire_lock_request.pb.go:1]: Does anyone external use AcquireLock?

**Review by snehagunta — APPROVED** [review]
I updated the branch with the latest authz refactor changes. This PR lgtm!

File: api/kessel/inventory/v1beta2/relation_tuple_filter.proto [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+// This type exists only for RBAC backward compatibility and will be removed.
+//
+// RelationTupleFilter is used to filter tuples based on their resource, relation, and subject.
+// All fields are optional but at least one field should be provided.
+message RelationTupleFilter {
+  optional string resource_namespace = 1;
+  optional string resource_type = 2;
+  optional string resource_id = 3;
+  optional string relation = 4;
+  optional RelationSubjectFilter subject_filter = 5;
```
[the remark this answers is not in the document]
Comment (merlante) [reply api/kessel/inventory/v1beta2/relation_tuple_filter.proto:21]: Empty filter: SpiceDB already validates this (`checkIfFilterIsEmpty in internal/services/v1/relationships.go:637`) and rejects empty filters. Relations-api doesn't validate server-side - it relies on SpiceDB. Adding validation would be an improvement (fail faster), but it's outside the scope of this PR which focuses on moving over endpoint "as is".                                                             
                                                            
I would be willing to add proto-level validation (via buf.validate annotations), where it doesn't introduce new failure modes, but making changes in the coverter code to add validation is outside of scope here.

File: cmd/serve/serve.go [hunk 4]
```diff
@@ -281,6 +296,15 @@ func NewCommand(
...
 
+			// DEPRECATED: Legacy tuple service for RBAC-only backward compatibility
+			tuple_crud_usecase := tuplesctl.New(
+				authorizer,
+				metaauthorizer.NewWhitelistMetaAuthorizer(metaAuthorizerConfig.TupleCrudAllowlist),
+				log.With(logger, "subsystem", "tuple_crud_controller"),
+			)
+			tuple_service := tuplesvc.New(tuple_crud_usecase)
+			pbv1beta2.RegisterKesselTupleServiceServer(server.GrpcServer, tuple_service)
+
```
[the remark this answers is not in the document]
Comment (merlante) [reply cmd/serve/serve.go:308]: Already implemented. All tuple CRUD endpoints enforce OIDC-only access via the WhitelistMetaAuthorizer:                                                                                                  
                                                                                                         
  - internal/biz/usecase/metaauthorizer/whitelist_metaauthorizer.go:38-41 explicitly checks AuthType == OIDC and denies non-OIDC requests                                                                  
  - All four operations (CreateTuples, DeleteTuples, ReadTuples, AcquireLock) call EnforceMetaAuthzObject with this meta authorizer (lines 35, 59, 84, 100 in tuple_crud_usecase.go)                       
                                                                                                                                                                                                           
  The WhitelistMetaAuthorizer enforces three requirements:                                                                                                                                                 
  1. Authenticated (line 29)                                                                                                                                                                               
  2. gRPC protocol only (line 34)                                                                                                                                                                          
  3. OIDC auth type only (line 39)                                                                                                                                                                         
                                                                                                                                                                                                           
  Non-OIDC requests (including x-rh-identity) are rejected before reaching the business logic.

File: internal/authz/simple_authorizer.go [hunk 5]
```diff
@@ -555,6 +555,46 @@ func matchesFilter(key tupleKey, filter *kessel.RelationTupleFilter) bool {
...
+							Id: key.SubjectID,
+						},
+					},
+				},
+			})
+		}
+	}
+
+	return &simpleReadTuplesStream{results: results}, nil
+}
```
[the remark this answers is not in the document]
Comment (merlante) [reply internal/authz/simple_authorizer.go]: Valid issue, but it applies only to a testing mock and is outside the scope of this PR.

File: internal/service/tuples/tuples.go [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+					Namespace: rel.GetSubject().GetSubject().GetType().GetNamespace(),
+					Name:      rel.GetSubject().GetSubject().GetType().GetName(),
+				},
+				Id: rel.GetSubject().GetSubject().GetId(),
+			},
+		},
+	}
+	// Set subject relation if present
+	if rel.GetSubject().Relation != nil {
+		result.Subject.Relation = rel.GetSubject().Relation
```
[the remark this answers is not in the document]
Comment (merlante) [reply internal/service/tuples/tuples.go:321]: The code is safe because proto validation ensures subject is required. Adding defensive nil checks would be redundant with the validation layer. The current pattern (check top-level nil at line 297, trust validation for required fields) is consistent with the rest of the function.

File: api/kessel/inventory/v1beta2/acquire_lock_request.pb.go (hunk 2 above)
Comment (merlante) [reply api/kessel/inventory/v1beta2/acquire_lock_request.pb.go:1]: Nobody external should be using any of these endpoints (and none have reported doing so).

File: api/kessel/inventory/v1beta2/acquire_lock_response.proto [hunk 7]
```diff
@@ -0,0 +1,13 @@
+syntax = "proto3";
+
+package kessel.inventory.v1beta2;
+
+option go_package = "github.com/project-kessel/inventory-api/api/kessel/inventory/v1beta2";
+option java_multiple_files = true;
+option java_package = "org.project_kessel.api.inventory.v1beta2";
+
+// DEPRECATED: AcquireLockResponse is part of the legacy tuple-layer API.
+// This type exists only for RBAC backward compatibility and will be removed.
```
Comment (tylercreller) [inline api/kessel/inventory/v1beta2/acquire_lock_response.proto:10]: AcuqireLock is also used for fencing Inventory consumer, such that in the event of a rebalance only one consumer will only ever acquire the lock. There are currently no plans to deprecate this as it was implemented for consistency reasons.

https://redhat.atlassian.net/browse/RHCLOUD-40702

edit: The same could be said for the Create/Delete tuples as well

File: .cursor/plans/relationsrepository_refactor_6ed96531.plan.md
Comment (tylercreller) [inline .cursor/plans/relationsrepository_refactor_6ed96531.plan.md:1]: Is this file intended to be merged as apart of this? It looks like a plan for a different effort via #1308

File: internal/biz/usecase/tuples/tuple_crud_usecase.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+		Authz:          authz,
+		MetaAuthorizer: metaAuthorizer,
+		Log:            log.NewHelper(logger),
+	}
+}
+
+// CreateTuples creates relationship tuples (DEPRECATED).
+// This endpoint exists only for RBAC backward compatibility.
+func (uc *TupleCrudUseCase) CreateTuples(ctx context.Context, cmd CreateTuplesCommand) (*CreateTuplesResult, error) {
+	uc.Log.Warn("DEPRECATED: CreateTuples called - this endpoint is for RBAC-only backward compatibility")
```
Comment (tylercreller) [inline internal/biz/usecase/tuples/tuple_crud_usecase.go:33]: nit: thoughts on lowering this (and others in this file) to `info` level? Warn could produce very noisy logs and would not be super helpful in our live envs.

File: .cursor/plans/relationsrepository_refactor_6ed96531.plan.md
Comment (merlante) [reply .cursor/plans/relationsrepository_refactor_6ed96531.plan.md:1]: I meant to ask @snehagunta the same thing. This is from her merge. :) Sneha?

File: internal/biz/usecase/tuples/tuple_crud_usecase.go (hunk 8 above)
Comment (merlante) [reply internal/biz/usecase/tuples/tuple_crud_usecase.go:33]: Is your concern that amount of logs generated could be more, or that the "warn" level could drown out other "warn" logging?

File: api/kessel/inventory/v1beta2/acquire_lock_response.proto (hunk 7 above)
Comment (merlante) [reply api/kessel/inventory/v1beta2/acquire_lock_response.proto:10]: This is the same use-case as for rbac. But re the inventory consumer, are you talking about the consumer in the inventory-api (this) repository, i.e. consumer.go? If so, we will not need (in the future) to expose proto externally for this. We are calling relations endpoints now, in the future, we will call an internal spicedb repository. Only in an interim period will we call these endpoints. So I think the inventory use-case will not extend the lifetime of these endpoints.

File: .cursor/plans/relationsrepository_refactor_6ed96531.plan.md
Comment (snehagunta) [reply .cursor/plans/relationsrepository_refactor_6ed96531.plan.md:1]: oops, my bad. removing.

File: internal/biz/usecase/tuples/tuple_crud_usecase.go (hunk 8 above)
Comment (tylercreller) [reply internal/biz/usecase/tuples/tuple_crud_usecase.go:33]: I think both points plus the added bit that I'm not sure the log provides us much benefit at that level.

File: api/kessel/inventory/v1beta2/acquire_lock_response.proto (hunk 7 above)
Comment (tylercreller) [reply api/kessel/inventory/v1beta2/acquire_lock_response.proto:10]: makes sense, thanks Mark!

File: internal/biz/usecase/tuples/tuple_crud_usecase.go (hunk 8 above)
Comment (Rajagopalan-Ranganathan) [reply internal/biz/usecase/tuples/tuple_crud_usecase.go:33]: +1 - Since we know it will be RBAC only, info level message would suffice and as we shouldnt encourage any other external service to use these end points. We have meta-authz for that

File: internal/biz/usecase/tuples/util.go [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+	if pagination.Continuation != nil {
+		req.ContinuationToken = pagination.Continuation
+	}
+	return req
+}
+
+// consistencyToV1beta1 converts model.Consistency to v1beta1 Consistency.
+// Note: Inventory has 3 consistency modes, relations-api has 2.
+// The at_least_as_acknowledged mode is mapped to minimize_latency as a safe downgrade.
+func consistencyToV1beta1(c model.Consistency) *relationspb.Consistency {
```
Comment (Rajagopalan-Ranganathan) [inline internal/biz/usecase/tuples/util.go:148]: I think this method is common for all the use case layer things, is there a common place we can place this? and these kind of things?

File: internal/biz/usecase/tuples/util.go [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+	}
+	return &relationspb.FencingCheck{
+		LockId:    fc.LockId,
+		LockToken: fc.LockToken,
+	}
+}
+
+// paginationToV1beta1 converts model.Pagination to v1beta1 RequestPagination.
+// Returns nil if pagination is not specified.
+func paginationToV1beta1(pagination *model.Pagination) *relationspb.RequestPagination {
```
Comment (Rajagopalan-Ranganathan) [inline internal/biz/usecase/tuples/util.go:132]: Same, common one wouldnt put it into a utils under tuple,  we need a common place for this and probably eliminate lot of duplication of code

File: internal/biz/usecase/tuples/tuple_crud_usecase.go (hunk 8 above)
Comment (merlante) [reply internal/biz/usecase/tuples/tuple_crud_usecase.go]: Done.

File: internal/biz/usecase/tuples/util.go (hunk 9 above)
Comment (merlante) [reply internal/biz/usecase/tuples/util.go:148]: @Rajagopalan-Ranganathan Take a look at Claude's analysis, below. Sounds like it's not really worth putting these functions in a common place. Only 2 overlapping functions, and one is different. Only one could be taken as truly common. We don't really want to change any existing implementation to suit the tuples crud usecase. Let me know what you think.

----
Analysis: Utility Function Duplication (Raj's PR Comment)

 The Question

 Raj noted that conversion functions in internal/biz/usecase/tuples/util.go may be duplicated elsewhere and should potentially be moved to a common location. Specifically:
 - consistencyToV1beta1
 - paginationToV1beta1

 Key question: Are these truly duplicated? Should they be consolidated into a shared package like internal/biz/usecase/common/?

 Findings

 1. paginationToV1beta1: EXACT DUPLICATE ✅

 tuples/util.go (lines 130-143):
 func paginationToV1beta1(pagination *model.Pagination) *relationspb.RequestPagination {
        if pagination == nil {
                return nil
        }
        req := &relationspb.RequestPagination{
                Limit: pagination.Limit,
        }
        if pagination.Continuation != nil {
                req.ContinuationToken = pagination.Continuation
        }
        return req
 }

 resources/resource_service.go (lines 818-828):
 func paginationToV1beta1(pagination *model.Pagination) *kessel.RequestPagination {
        if pagination == nil {
                return nil
        }
        result := &kessel.RequestPagination{
                Limit: pagination.Limit,
        }
        if pagination.Continuation != nil {
                result.ContinuationToken = pagination.Continuation
        }
        return result
 }

 Status: Exact duplicate - only difference is variable naming (req vs result) and proto package alias (relationspb vs kessel). Both import the same underlying proto:
 github.com/project-kessel/relations-api/api/kessel/relations/v1beta1

 Consolidation: Strong candidate for deduplication.

 ---
 2. consistencyToV1beta1: DIFFERENT IMPLEMENTATIONS ⚠️

 tuples/util.go (lines 145-174): Comprehensive implementation handling all 3 consistency modes
 func consistencyToV1beta1(c model.Consistency) *relationspb.Consistency {
        switch model.ConsistencyTypeOf(c) {
        case model.ConsistencyMinimizeLatency:
                return &relationspb.Consistency{
                        Requirement: &relationspb.Consistency_MinimizeLatency{MinimizeLatency: true},
                }
        case model.ConsistencyAtLeastAsFresh:
                fresh, _ := model.AsAtLeastAsFresh(c)
                return &relationspb.Consistency{
                        Requirement: &relationspb.Consistency_AtLeastAsFresh{
                                AtLeastAsFresh: &relationspb.ConsistencyToken{
                                        Token: fresh.ConsistencyToken().Serialize(),
                                },
                        },
                }
        case model.ConsistencyAtLeastAsAcknowledged:
                // Relations-api doesn't support at_least_as_acknowledged - default to minimize_latency
                return &relationspb.Consistency{
                        Requirement: &relationspb.Consistency_MinimizeLatency{MinimizeLatency: true},
                }
        default:
                // Unspecified - default to minimize_latency
                return &relationspb.Consistency{
                        Requirement: &relationspb.Consistency_MinimizeLatency{MinimizeLatency: true},
                }
        }
 }

 resources/resource_service.go (lines 833-848, 864-879): Inline implementation in two places, only handles 2 consistency modes
 // In lookupResourcesCommandToV1beta1 and lookupSubjectsCommandToV1beta1:
 var consistency *kessel.Consistency
 if token := model.ConsistencyAtLeastAsFreshToken(cmd.Consistency); token != nil {
        consistency = &kessel.Consistency{
                Requirement: &kessel.Consistency_AtLeastAsFresh{
                        AtLeastAsFresh: &kessel.ConsistencyToken{
                                Token: token.Serialize(),
                        },
                },
        }
 } else {
        consistency = &kessel.Consistency{
                Requirement: &kessel.Consistency_MinimizeLatency{
                        MinimizeLatency: true,
                },
        }
 }

 Key differences:
 - tuples/ - Comprehensive, handles all 3 consistency types (MinimizeLatency, AtLeastAsFresh, AtLeastAsAcknowledged) with explicit downgrade for unsupported type
 - resources/ - Simpler, only handles 2 types (AtLeastAsFresh vs default MinimizeLatency)
 - resources/ explicitly rejects AtLeastAsAcknowledged earlier in the call chain (lines 399, 423, 486, 504) with error: "inventory-managed consistency tokens aren't available"

 Status: Different implementations with different assumptions. Resources usecase doesn't need comprehensive handling because it fails-fast on unsupported consistency types.

 Consolidation: Possible, but would require either:
 1. Using tuple version (comprehensive) everywhere, OR
 2. Creating shared version that matches tuple behavior
 3. Ensuring resources/ still validates/rejects AtLeastAsAcknowledged appropriately

 ---
 Where Should Shared Code Go?

 1. internal/biz/usecase/common/ (new package)

 ---
 Risks of Consolidation

 1. Import cycles: Need to verify that moving to a new common package won't create circular dependencies
 2. Premature abstraction: Only 2 usecases currently (tuples, resources). Future usecases might not need these exact converters. Rule of three suggests waiting until 3 instances before extracting.
 3. Behavioral divergence: The consistency handling differs intentionally between usecases. Consolidating requires careful analysis of whether resources/ should adopt tuple behavior or maintain its own
 validation.
 4. Churn for deprecated code: Tuple CRUD endpoints are deprecated and exist only for RBAC backward compatibility. If they're removed soon (post-RBAC migration), consolidation effort may not be
 worthwhile.
 5. Test coverage: Both implementations have tests. Consolidation requires ensuring all edge cases are still covered.

 ---
 Recommendation

 paginationToV1beta1: ✅ Consolidate
 - Clear duplication with zero semantic differences
 - Low risk, high value
 - Extract to internal/biz/usecase/common/conversions.go

 consistencyToV1beta1: ❌ Don't consolidate (yet)
 - Different implementations with different assumptions
 - Resources usecase intentionally doesn't handle AtLeastAsAcknowledged
 - Consolidating would require changing resources/ behavior or adding complexity
 - Tuple endpoints are deprecated - may be removed before consolidation pays off

 Other functions in tuples/util.go:
 - Tuple-specific converters (relationsTupleToV1beta1, tupleFilterToV1beta1, etc.) - no duplication found, keep in tuples/

 ---

Comment (merlante) [reply internal/biz/usecase/tuples/util.go:148]: Going to just leave as is, as there's not much benefit to moving just 1 or 2 functions to a common place.

File: internal/biz/usecase/tuples/util.go (hunk 10 above)
Comment (merlante) [reply internal/biz/usecase/tuples/util.go:132]: See above comment.

File: internal/biz/usecase/tuples/util.go (hunk 9 above)
Comment (Rajagopalan-Ranganathan) [reply internal/biz/usecase/tuples/util.go:148]: we can leave this for now based on the analysis and can refactor again if needed in the future

File: internal/biz/usecase/tuples/util.go (hunk 10 above)
Comment (Rajagopalan-Ranganathan) [reply internal/biz/usecase/tuples/util.go:132]: see above comment.

**Review by Rajagopalan-Ranganathan — APPROVED** [review]
